exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,452 @@
+            let view_handle = to_handle(View::from_ref_snapshot(snapshot));
+            let java_runtime_hashes = panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::state_hashes_id(),
+                    JavaType::Array(Box::new(JavaType::Primitive(Primitive::Byte))),
+                    &[JValue::from(view_handle)],
+                ),
+            );
+            let byte_array = java_runtime_hashes.l()?.into_inner();
+            let data = env.convert_byte_array(byte_array)?;
+
+            Ok(data)
+        }));
+
+        ServiceRuntimeStateHashes::from_bytes(bytes.into())
+            .unwrap()
+            .into()
+    }
+
+    fn before_commit(&self, _dispatcher: &DispatcherRef, fork: &mut Fork) {
+        unwrap_jni(self.exec.with_attached(|env| {
+            let view_handle = to_handle(View::from_ref_fork(fork));
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::before_commit_id(),
+                    JavaType::Primitive(Primitive::Void),
+                    &[
+                        JValue::from(view_handle),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+
+    fn after_commit(
+        &self,
+        _dispatcher: &DispatcherSender,
+        snapshot: &Snapshot,
+        service_keypair: &(PublicKey, SecretKey),
+        _tx_sender: &ApiSender,
+    ) {
+        unwrap_jni(self.exec.with_attached(|env| {
+            let view_handle = to_handle(View::from_ref_snapshot(snapshot));
+            let validator_id = Self::validator_id_or(snapshot, &service_keypair.0, -1);
+            let height:u64 = CoreSchema::new(snapshot).height().into();
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::after_commit_id(),
+                    JavaType::Primitive(Primitive::Void),
+                    &[
+                        JValue::from(view_handle),
+                        JValue::from(validator_id),
+                        JValue::from(height as i64),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+
+    fn api_endpoints(&self, context: &ApiContext) -> Vec<(String, ServiceApiBuilder)> {
+        Vec::new()
+    }
+
+    fn notify_api_changes(&self, context: &ApiContext, changes: &[ApiChange]) {
+        let added_instances_ids: Vec<i32> = changes.iter()
+            .filter_map(|change| {
+                if let ApiChange::InstanceAdded(instance_id) = change {
+                    Some(*instance_id as i32)
+                } else {
+                    None
+                }
+            })
+            .collect();
```

Comment:
Please see the docs of the runtime method: added_instances_ids must not be empty (connect_apis shall not be invoked if no new services need to be connected). That is a temporary requirement till proper Node handling is implemented.